py/server/tests/test_table_listener.py

```diff
@@ -3,101 +3,162 @@
-
 from tests.testbase import BaseTestCase
 
 
 class TableListenerTestCase(BaseTestCase):
+
     def setUp(self) -> None:
         with self.ugp_lock_exclusive():
-            self.table = time_table("00:00:01").update(["X=i"]).tail(5)
-        self.table_change_processed = None
-
-    def test_shift_oblivious_listener_obj(self):
+            self.test_table = time_table("00:00:00.001").update(["X=i%2"]).sort("X").tail(16)
+            source_table = time_table("00:00:00.001").update(["TS=currentTime()"])
+            self.test_table2 = time_window(source_table, ts_col="TS", window=10 ** 7, bool_col="InWindow")
+
+    def tearDown(self) -> None:
+        self.test_table = None
+        self.test_table2 = None
+
+    def check_result(self, changes: List[Dict[str, numpy.ndarray]], cols: Union[str, List[str]] = None):
+        if not changes:
+            return
+        cols = to_sequence(cols)
+        for change in changes:
+            self.assertTrue(isinstance(change, dict))
+            if not cols:
+                cols = [col.name for col in self.test_table.columns]
+            for col in cols:
+                self.assertIn(col, change.keys())
+                self.assertTrue(isinstance(change[col], numpy.ndarray))
+                self.assertEqual(change[col].ndim, 1)
+                # print(change[col])
+
+    def test_listener_obj(self):
         class ListenerClass(TableListener):
             def __init__(self):
-                self.table_change_added = None
-                self.table_change_removed = None
-                self.table_change_modified = None
-
-            def onUpdate(self, added, removed, modified):
-                self.table_change_added = added
-                self.table_change_removed = removed
-                self.table_change_modified = modified
-
-            def has_table_changed(self):
-                return (
-                    self.table_change_removed
-                    or self.table_change_modified
-                    or self.table_change_added
-                )
+                self.table_update = SimpleNamespace(added=None, removed=None, modified=None, modified_prev=None,
+                                                    Shifted=None)
+
+            def onUpdate(self, is_replay, update):
+                update.chunk_size = 4
+                self.table_update.added = []
+                for chunk in update.added("X"):
+                    self.table_update.added.append(chunk)
+
+                self.table_update.removed = []
+                for chunk in update.removed():
+                    self.table_update.removed.append(chunk)
+
+                self.table_update.modified = []
+                for chunk in update.modified():
+                    self.table_update.modified.append(chunk)
+
+                self.table_update.modified_prev = []
+                for chunk in update.modified_prev():
+                    self.table_update.modified_prev.append(chunk)
```

Comment:
A user may want to look at differences between the modified and modified_prev values.  Since chunking and generators are being used, what would expect the design pattern to be like?  Something like below?

```
for m, mp in zip(update.modified(), update.modified_prev()):
    ... do stuff ...
```